risc0/zkvm/src/prove/image.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::{bail, Result};
+use risc0_zkp::core::sha::{Digest, Sha};
+use risc0_zkvm_platform::{
+    memory::{MEM_SIZE, PAGE_TABLE, PROG},
+    syscall::DIGEST_BYTES,
+    WORD_SIZE,
+};
+
+use crate::{prove::elf::Program, sha};
+
+struct PageTableInfo {
+    page_size: usize,
+    layers: Vec<usize>,
+    total: usize,
+}
+
+impl PageTableInfo {
+    pub fn new(max_mem: usize, page_size: usize) -> Self {
+        assert!(max_mem >= page_size);
+
+        let mut layers = Vec::new();
+        let mut total = 0;
+        let mut remain = max_mem;
+        while remain >= page_size {
+            let num_pages = remain / page_size;
+            remain = num_pages * DIGEST_BYTES;
+            layers.push(remain);
+            total += remain;
+        }
+
+        Self {
+            page_size,
+            layers,
+            total,
+        }
+    }
+}
+
+pub struct MemoryImage {
+    pub image: Vec<u8>,
+    pub root: Digest,
+    info: PageTableInfo,
+}
+
+impl MemoryImage {
+    pub fn new(elf_contents: &[u8], page_size: usize) -> Result<Self> {
```

Comment:
Why not return `Self` instead of `Result<Self>` on this function?

Reply:
Because `Program::load_elf` can return an error, and we need to avoid panics.